docs/tutorials/interceptors.md

```diff
@@ -0,0 +1,123 @@
+Interceptors in AI DIAL could be classified into the following categories:
+
+* **Pre-interceptors** that only modify the incoming request from the client (e.g. rejecting requests following certain criteria)
+* **Post-interceptors** that only modify the response received from the upstream (e.g. censoring the response)
+* **Generic interceptors** that modify both the incoming request and the response from the upstream (e.g. caching the responses)
+
+Interceptors in AI DIAL can:
+
+* Modify an incoming AI DIAL request received from the client (or it may leave it as is).
+* Call an upstream AI DIAL application (the upstream for short) with the modified request.
+* Modify the response from the upstream (or it may leave it as is).
+* Return the modified response to the client.
+
+For example, interceptors can block requests that violate specific regulations, related with restricted domains, or potentially lead to data leaks or biased responses. Another use case is when interceptors allow applications or models to respond solely to specific subjects and anonymize Personally Identifiable Information (PII) from user requests, or cache LLM responses.
+
+To implement PII (Personally Identifiable Information) anonymization for all data sent to models through AI DIAL, you can use interceptors which can employ specific locally deployed NLP models to obfuscate (replace with token) PII in requests (pre-interceptor) and decode it in responses (post-interceptor), effectively ensuring the anonymization of all personal data.
+
+## Flow
+
+Technically speaking, interceptors in AI DIAL are components inserted into deployments (applications or model adapters) that can be called before or after [chat completion requests](https://epam-rail.com/dial_api#/paths/~1openai~1deployments~1%7BDeployment%20Name%7D~1chat~1completions/post).
+
+The stack of interceptors in AI DIAL works similarly to a stack of middleware in Express.js or Django. Every request/response in the diagram above goes through AI DIAL Core (this is hidden from the diagram for brevity):
+
+```js
+Client -> (original request) ->
+  Interceptor 1 -> (modified request #1) ->
+    Interceptor 2 -> (modified request #2) ->
+      Upstream -> (original response) ->
+    Interceptor 2 -> (modified response #1) ->
+  Interceptor 1 -> (modified response #2) ->
+Client
+```
+
+AI DIAL Core manages chat completion requests from interceptors through the endpoint: `/openai/deployments/<deployment|interceptor>/chat/completions`. It uses the deployment name `interceptor` to handle requests from all interceptors. Upon receiving a request, it identifies the next interceptor based on the current interceptor specified in the API Key data. The final interceptor in the sequence is always the target deployment (application, model).
+
+To demonstrate the flow on a more specific example, lets take two interceptors **gtp-cache** and **pii-anonymizer** configured for the **GPT-4** model:
+
+```json
+//configuration in AI DIAL Core
```

Comment:
Here it's called `configuration in AI DIAL Core`, later the **same** config is called `AI DIAL Core dynamic settings`. Let's make it consistent.

Reply:
fixed